ncm-shorewall/src/main/perl/shorewall.pm

```diff
@@ -1,436 +1,170 @@
-                                    path
-                                    shorewall_shell
-                                    subsyslock
-                                    modulesdir
-                                    config_path
-                                    restorefile
-                                    ipsecfile
-                                    lockfile
-                                    drop_default
-                                    reject_default
-                                    accept_default
-                                    queue_default
-                                    nfqueue_default
-                                    rsh_command
-                                    rcp_command
-                                    ip_forwarding
-                                    tc_enabled
-                                    tc_priomap
-                                    mutex_timeout
-                                    module_suffix
-                                    maclist_table
-                                    maclist_ttl
-                                    optimize
-                                    dont_load
-                                    zone2zone
-                                    forward_clear_mark
-                                    blacklist_disposition
-                                    maclist_disposition
-                                    tcp_flags_disposition
-                                    );
+Readonly my $CONFIG_DIR => "/etc/shorewall";
+Readonly my $BACKUP_SUFFIX => '.quattor.';
+Readonly my $FAILED_SUFFIX => '.failed.';
+Readonly my $SHOREWALL_CHECK => ['/sbin/shorewall', 'check', $CONFIG_DIR];
+Readonly my $SHOREWALL_TRY => ['/sbin/shorewall', 'try', $CONFIG_DIR];
+Readonly my $CCM_FETCH => [qw(/usr/sbin/ccm-fetch)];
 
-sub writecfg
+# Generate new config files, keep backups
+# Return undef on failure, the filename if file was modified, or 0 otherwise.
+sub make_config
 {
-    my ($self, $typ, $contents, $refreload) = @_;
-
-    my $changed = 0;
-    my $cfgfile=SHOREWALLCFGDIR.$typ;
-
-    $cfgfile .=".conf" if ($typ eq "shorewall");
-
-    if ( -e $cfgfile && (LC::File::file_contents($cfgfile) eq $contents)) {
-        $self->debug(2,"Nothing changed for $typ.")
-    } else {
-        ## write cfgfile
-        $changed = 1;
-        my $result = LC::Check::file( $cfgfile,
-                                   backup => ".old",
-                                   contents => encode_utf8($contents),
-                                  );
-        if ($result) {
-            $self->info("$cfgfile updated");
-        } else {
-            $self->error("$cfgfile update failed");
-            $changed=-1;
-        }
+    my ($self, $feat, $config, $ts) = @_;
+
+    my $filename = "$CONFIG_DIR/$feat";
+    $filename .= '.conf' if $feat eq 'shorewall';
+
+    my $trd = EDG::WP4::CCM::TextRender->new(
+        $feat,
+        $config->getElement($self->prefix()."/$feat"),
```

Comment:
@piojo i'm not sure what was discussed "yesterday", but passing an element instance and a hash are not the same thing in `CCM::TextRender`. passing an element allows access to type information.
this is probably the only reason why component authors should access the config instance instead of the `->getTree` result.